compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java

```diff
@@ -120,4 +135,114 @@ public Spliterator spliterator() {
             return this.documentList.spliterator();
         }
     }
+
+    /**
+     * Inner class that handles module modifications.
+     */
+    public static class Modifier {
+        private ModuleId moduleId;
+        private ModuleName moduleName;
+        private Map<DocumentId, DocumentContext> srcDocContextMap;
+        private Map<DocumentId, DocumentContext> testDocContextMap;
+        private boolean isDefaultModule;
+        private Set<ModuleDependency> moduleDependencies;
+        private Package packageInstance;
+        private Project project;
+
+
+        private Modifier(Module oldModule) {
+            moduleId = oldModule.moduleId();
+            moduleName = oldModule.moduleName();
+            srcDocContextMap = copySrcDocs(oldModule);
+            testDocContextMap = copyTestDocs(oldModule);
+            isDefaultModule = oldModule.isDefaultModule();
+            moduleDependencies = new HashSet<>(oldModule.moduleDependencies());
+            packageInstance = oldModule.packageInstance;
+            project = oldModule.project();
+        }
+
+        Modifier updateDocument(DocumentContext newDocContext) {
+            if (this.srcDocContextMap.get(newDocContext.documentId()) != null) {
+                this.srcDocContextMap.put(newDocContext.documentId(), newDocContext);
+            } else {
+                this.testDocContextMap.put(newDocContext.documentId(), newDocContext);
+            }
+            return this;
+        }
+
+        /**
+         * Creates a copy of the existing module and adds a new source document to the new module.
+         *
+         * @param documentConfig configurations to create the document
+         * @return an instance of the Module.Modifier
+         */
+        public Modifier addDocument(DocumentConfig documentConfig) {
+            DocumentContext newDocumentContext = DocumentContext.from(documentConfig);
+            this.srcDocContextMap.put(newDocumentContext.documentId(), newDocumentContext);
+            return this;
+        }
+
+        /**
+         * Creates a copy of the existing module and adds a new test document to the new module.
+         *
+         * @param documentConfig configurations to create the document
+         * @return an instance of the Module.Modifier
+         */
+        public Modifier addTestDocument(DocumentConfig documentConfig) {
+            DocumentContext newDocumentContext = DocumentContext.from(documentConfig);
+            this.testDocContextMap.put(newDocumentContext.documentId(), newDocumentContext);
+            return this;
+        }
+
+        /**
+         * Creates a copy of the existing module and removes the specified document from the new module.
+         *
+         * @param documentId documentId of the document to remove
+         * @return an instance of the Module.Modifier
+         */
+        public Modifier removeDocument(DocumentId documentId) {
+
+            if (this.srcDocContextMap.get(documentId) != null) {
```

Comment:
If we do not allow null values in the map, let's use the `Map.containsKey` API

Reply:
Fixed in 0d89c8f